### PR TITLE
 Consistent evaluation and display of validation results

### PR DESCRIPTION
**Pull request by nspool**
Fixes bugs and limitations in validation to allow rules to be evaluated more consistently.

 - Change validations to remove use of Tuples and deep nested maps
 - Move most validations into methods within ValidateProcess
 - Use the ModelMapper instead of hashmap to map between StagedRow and 
   StagedRowFormatted
 - Change and simplify JSON responses to easier display validations in UI
 - Replace validation UI accordion with a tree list
 - Add validation highlights to allow adjacent cells related to a
   single validation rule to be highlighted together
 - Begin updating unit tests

Add a sheet summary in the validation panel and in the job log on ingest with a count of data rows, distinct sites, distinct surveys, distinct observable items and distinct divers. Also on the validation panel it displays the number of incomplete surveys.

### CODE REVIEW

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/process/ValidationProcess.java [hunk 1]
```diff
@@ -1,117 +1,525 @@
...
+            List<StagedRowFormatted> surveyRows = survey.getValue();
+
+            // VALIDATE: Survey Group Complete
+            res.add(validateSurveyGroup(survey.getKey(), surveyRows));
+
+            // VALIDATE: Is Existing Survey
+            res.add(validateSurveyIsNew(surveyRows.get(0)));
+
+            // VALIDATE: Has M1 M2
+            if (programName == "RLS" || !validateATRCM3457(surveyRows))
```
Comment (APaulic) [inline api/src/main/java/au/org/aodn/nrmn/restapi/validation/process/ValidationProcess.java:442]: String comparison, so it should be `programName.equals("RLS")`

File: ui/src/components/import/panel/ValidationSummary.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+              </Typography>
+              <Typography variant="caption" color="inherit">
+                {`(${m.description.length})`}
+              </Typography>
+            </div>
+          }
+          key={m.message}
+        >
+          {m.description.map((d) => {
+            const mmHeader = mm.find((m) => m.field === d.columnName);
```
Comment (APaulic) [inline ui/src/components/import/panel/ValidationSummary.js:63]: Another string comparison `m.field.equals(d.columnName)`

**Review by APaulic — CHANGES_REQUESTED** [review]
The listed bugs and just the string comparisons to avoid strange behaviour later.

**Comment by bpasquer** [comment]
Feedback from L5-L95 checks successfully group error by species name. 
- For single observations outside the valid range in a record(site/date/depth/method/block), the cell containing the observation is highlighted and  the message indicates the value(size) and number of observations of this size for the particular species for ex: `40/5.5 2` ( 40 FishSize / 5.5 Invertsize) 2 observations
- For multiple observations outside the valid range in a record(site/date/depth/method/block), the message says:`Region contains invalid values` and the region spanning the invalid observations is highlighted.

However, 
- the highlighted region sometimes extends over empty cells (see feedback for Olisthops cyanomelas)
- the highlighted region contains a single value (see Cheilodactylus spectabilis)
- Feedback for single observations shows invalid range in a single record (see Acanthaluteres vittiger)

[Bicheno_clean.xlsx](https://github.com/aodn/nrmn-application/files/6860184/Bicheno_clean.xlsx)

File: ui/src/components/import/panel/ValidationSummary.js (hunk 2 above)
Comment (nspool) [reply ui/src/components/import/panel/ValidationSummary.js:63]: This is OK for javascript.

File: api/src/main/java/au/org/aodn/nrmn/restapi/validation/process/ValidationProcess.java (hunk 1 above)
Comment (nspool) [reply api/src/main/java/au/org/aodn/nrmn/restapi/validation/process/ValidationProcess.java:442]: Fixed.

**Comment by nspool** [comment]
@bpasquer fixed an issue where highlights were appearing on multiple rows for some L5-L95 validation.

Running up another stack at http://nrmn-refactor-validation.dev.aodn.org.au/

**Comment by nspool** [comment]
Also completing https://github.com/aodn/backlog/issues/3135 as part of this issue. 
Generating the same data that the RLS tool produces, eg:

2480 rows of data found
17 distinct sites found
34 distinct surveys found
	0 incomplete
231 distinct observable items found
7 distinct divers found

Displaying in both the validation panel and in the job log on ingest.

**Pull request merged**